pandas/core/frame.py

```diff
@@ -8359,6 +8359,19 @@ def groupby(
     ) -> DataFrameGroupBy:
         from pandas.core.groupby.generic import DataFrameGroupBy
 
+        if isinstance(by, list):
+            if len(by) == 1 and isinstance(by[0], str):
+                warnings.warn(
+                    (
+                        "In a future version of pandas, a length 1 "
+                        "tuple will be returned when grouping by a "
+                        "list of length 1. Don't supply a list with "
+                        "a single grouper to avoid this warning."
+                    ),
+                    FutureWarning,
+                    stacklevel=find_stack_level(),
+                )
```

Comment:
We should only be warning when the user will experience a behavior change once the deprecation is enforced. I believe the only difference will be when a user iterates over a groupby as in `for key, group in df.groupby(["a"]):`, is that right? We should then only warn when iterating.

Reply:
@rhshadrach 

The thing is that the `groupby` object will change by changing one property inside it. Also, iterating over `groupby` is done by other functions like plotting functions and the pivot table. I believe that the user should know that in general, they should pass the key without a list as long as they don't want the key to be in form of a tuple.

Reply:
> The thing is that the groupby object will change by changing one property inside it.

I don't think this is important. It's only important when the results change to a user.

> Also, iterating over groupby is done by other functions like plotting functions and the pivot table.

If this is all handled internally by pandas, then the user does not need to know about it. Only when the results that the user sees will change do they need to be warned.

> I believe that the user should know that in general, they should pass the key without a list as long as they don't want the key to be in form of a tuple.

If their results will not change, we do no want to warn them. Warning users unnecessarily makes them do work to avoid the warning and that is a bad user experience. Users are informed of general changes to pandas through our release notes (the whatsnew).